fix missing prevout tx fee underflow

mononaut opened the pull request:
Fixes an issue where transactions returned by `mempool/txs` endpoints could have underflowed fees due to missing prevouts.

`mempool.lookup_txos()` is not guaranteed to find all txos for mempool transactions. When we later use the result of `mempool.lookup_txos()` to create new TransactionValue objects, some prevout values may be missing causing incorrect fees to be calculated.

If the missing prevouts cause the sum of output values to be greater than the sum of the (remaining) inputs, the resulting u64 fee underflows, producing absurdly high fees.

This PR ~~adds a `filter_missing_prevouts` flag to `prepare_txs`, which~~ makes `TransactionValue::new` return an error instead of a bad result if any prevouts cannot be found. Any failures are then filtered out of the final list of prepared transactions.

mononaut committed:
Filter out transactions with missing prevouts from instead of returning bad data

junderw commented:
This seems like it should work.

But tbh this bool flag is pretty important and it's not documented.

What cases should one use true? When is it ok to use false? etc.

That should be readily apparent in the code, but since it's not, and it's hard to see the invariants that need to be upheld there should be a short explanation.

I also wonder if the address/.../txs endpoint should be true... Does the address page not deal with the prevouts?

junderw commented:
The /chain endpoint is all confirmed so false is fine (there should be none)

mononaut commented:
> But tbh this bool flag is pretty important and it's not documented.
>
> What cases should one use true? When is it ok to use false? etc.
>
> That should be readily apparent in the code, but since it's not, and it's hard to see the invariants that need to be upheld there should be a short explanation.

yup, I'll add some explanatory comments.

mononaut commented:
> I also wonder if the address/.../txs endpoint should be true... Does the address page not deal with the prevouts?

yes, I think we probably do want to filter out unconfirmed transactions with missing prevouts anywhere that's possible, including `/address/.../txs` etc

but I don't really understand why `allow_missing` is set to `true` here in the first place, so I want to be as conservative as possible with this change.
https://github.com/mempool/electrs/blob/a19ce10c43daf81bde7c36e84464cb92d4f572bf/src/rest.rs#L159-L160

mononaut committed:
Document filter_missing_prevouts flag

junderw commented:
> but I don't really understand why allow_missing is set to true here in the first place, so I want to be as conservative as possible with this change.

I remember writing that flag. It was to maintain a specific behavior in specific cases. (I think originally it was panicking)

I think it might be worth it to just remove transactions with missing outpoints from the Mempool cache................. but without knowing everything that triggers the missing Outpoints, it's kind of scary...

However, each loop of the main loop will check bitcoin core's mempool for new transactions, so maybe it's not that bad.

At least I think we can safely assume that we ALWAYS want to filter out no-prevout transactions......... maybe....

mononaut commented:
> I think it might be worth it to just remove transactions with missing outpoints from the Mempool cache

yeah, this feels like the correct long-term fix, but it isn't clear to me what the side effects could be.

I'd prefer to start with this limited PR for now to fix the embarrassing bug, and follow up with a long term solution after more thorough investigation.

junderw commented:
> I'd prefer to start with this limited PR for now to fix the embarrassing bug, and follow up with a long term solution after more thorough investigation.

Ok, then at least get rid of the boolean. We should always filter out the ones without parent outpoints for all endpoints IMO.

mononaut committed:
Remove filter_missing_prevout flag, always filter txs with bad prevouts

mononaut commented:
rebased, and removed the `filter_missing_prevout` and `allow_missing_prevout` flags.

Now `TransactionValue::new()` always throws an error if any prevouts are missing, and `prepare_txs()` always filters out those failed transactions.

junderw commented on src/rest.rs:
```diff
@@ -519,6 +519,15 @@ fn ttl_by_depth(height: Option<usize>, query: &Query) -> u32 {
+///
+/// Any transactions with missing prevouts will be filtered out of the response, rather than returned with incorrect data.
+///
+/// ## Arguments
+///
+/// * `txs`
+/// * `query`
+/// * `config`
```
nit: Not very helpful... incomplete? I don't think it's needed tbh.

junderw commented on src/rest.rs:
```diff
@@ -1003,7 +1012,15 @@ fn handle_request(
             let mut tx = prepare_txs(vec![(tx, blockid)], query, config);
 
-            json_response(tx.remove(0), ttl)
+            if tx.is_empty() {
+                http_message(
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    "Transaction missing prevouts",
+                    0,
+                )
+            } else {
+                json_response(tx.remove(0), ttl)
+            }
```
nit: Looks ugly............... but oh well.

junderw reviewed (approved):
LGTM, utACK

junderw committed:
Remove unneeded comments

junderw committed:
Remove allow_missing and fix clippy errors

junderw committed:
Fail CI when clippy gives warnings

junderw committed:
Fix error when liquid + test are enabled. Hide .vscode folder

junderw committed:
Run cargo tests as part of CI

wiz committed:
Merge branch 'mempool' into mononaut/fix-fee-underflow

wiz reviewed (approved):
Not sure how to test this but ran it for 10 minutes and it didn't blow up so 🤷 
Tested ACK @ 3.0.0-dev-7a5a27c

softsimon reviewed (approved):
tested ACK @ [7a5a27cbcaf]

Seems to be working
